### PR TITLE
Fix SpotBugs false positive in FilePath.IsDescendant

### PR DESCRIPTION
**Pull request by Vinamra-tech**
SpotBugs flagged a potential `NullPointerException` in `hudson.FilePath$IsDescendant.invoke` because `getDirectChild` was annotated with `@CheckForNull`. 

**Investigation findings:**
* **Dead Code Analysis:** The code path leading to a null return from `getDirectChild` is effectively unreachable in this context because the method handles traversal logic internally.
* **Root Cause:** The method `getDirectChild` was unnecessarily annotated with `@CheckForNull` relative to its usage here.

**Changes in this PR:**
* Updated the annotation on `getDirectChild` from `@CheckForNull` to `@NonNull`.
* Added a defensive `IllegalStateException` to satisfy the `@NonNull` contract and ensure safe failure if a regression ever exposes this path.

### Testing done
* **Regression Testing:** Verified `FilePathTest` passes.
* **Static Analysis:** Ran `mvn compile spotbugs:spotbugs -pl core` to confirm the warning is resolved.
* **Manual Verification:** Verified via Reflection that the defensive `IllegalStateException` is correctly thrown if the unreachable path is forced (see PR comments for reproduction).

### Proposed changelog entries
* N/A (Internal housekeeping/Static analysis fix)

### Proposed upgrade guidelines
N/A

### Submitter checklist

- [x] The issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-ci/jenkins/blob/master/CONTRIBUTING.md#changelog-entries)).
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since` TODO Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] UI changes do not introduce regressions when enforcing the current default rules of Content Security Policy Plugin.
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.

### Maintainer checklist

- [ ] The description looks good.
- [ ] The build passed.
- [ ] The changes look good.

### CODE REVIEW

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
Please use the pull request template, including its headings.  When you submit a pull request that does not complete the pull request template, you cause maintainers to distrust your pull request even before they read the source code.

**Comment by MarkEWaite** [comment]
Please correct the pull request title.  Since you have not provided any test that shows a null pointer exception is possible, you aren't fixing a null pointer exception.  At best you are preventing a possible null pointer exception.

Even better would be for you to provide an automated test that illustrates the null pointer exception and then the code change confirms that the null pointer exception is fixed.

**Comment by Vinamra-tech** [comment]
@MarkEWaite Thank you for the feedback. I apologize for the template issues; this is my first contribution to Jenkins core.

I have updated the Title to "Prevent potential..." to reflect that this is a proactive fix based on Static Analysis.

I have restored the Description Template fully.

regarding the test: Triggering Util.fileToPath to return null requires inducing an InvalidPathException (e.g., creating a file with illegal characters for the specific OS). If you have a recommendation for an existing test pattern in the codebase that mocks filesystem failures safely, I would be happy to implement it. Otherwise, I relied on SpotBugs verification as the logic fix is defensive.

File: core/src/test/java/hudson/FilePathTest.java [hunk 1]
```diff
@@ -1297,4 +1297,17 @@ private static File newFolder(File root, String... subDirs) throws IOException {
         }
         return result;
     }
+
+    @Test
+    @Issue("SpotBugs")
+    public void testIsDescendantWithParentTraversal() throws Exception {
+        File tmp = Util.createTempDir();
```
Comment (MarkEWaite) [inline core/src/test/java/hudson/FilePathTest.java]: Please use the JUnit `@TempDir` annotation to allow the JUnit framework to perform the directory cleanup rather than using try / finally to perform the cleanup yourself.  There are several examples available in the Jenkins tests.

**Review by MarkEWaite — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/FilePath.java [hunk 2]
```diff
@@ -3781,6 +3781,11 @@ public Boolean invoke(@NonNull File parentFile, @NonNull VirtualChannel channel)
             Path currentFilePath = parentFile.toPath();
             while (!remainingPath.isEmpty()) {
                 Path directChild = this.getDirectChild(currentFilePath, remainingPath);
+                // --- FIX #2: Handle if directChild is null ---
```
Comment (MarkEWaite) [inline core/src/main/java/hudson/FilePath.java]: The extra comment text is incorrect now and doesn't help comprehension
```suggestion
```

File: core/src/main/java/hudson/FilePath.java [hunk 3]
```diff
@@ -3781,6 +3781,11 @@ public Boolean invoke(@NonNull File parentFile, @NonNull VirtualChannel channel)
...
+                if (directChild == null) {
+                    return false;
+                }
+                // ---------------------------------------------
```
Comment (MarkEWaite) [inline core/src/main/java/hudson/FilePath.java]: Doesn't help comprehension of the change
```suggestion
```

File: core/src/test/java/hudson/FilePathTest.java [hunk 4]
```diff
@@ -1297,4 +1297,17 @@ private static File newFolder(File root, String... subDirs) throws IOException {
         }
         return result;
     }
+
+    @Test
+    @Issue("SpotBugs")
```
Comment (MarkEWaite) [inline core/src/test/java/hudson/FilePathTest.java]: We use the `@Issue` annotation for bug reports.  No need to annotate a test for a null pointer exception
```suggestion
```

File: core/src/test/java/hudson/FilePathTest.java [hunk 5]
```diff
@@ -1297,4 +1297,17 @@ private static File newFolder(File root, String... subDirs) throws IOException {
...
+    public void testIsDescendantWithParentTraversal() throws Exception {
+        File tmp = Util.createTempDir();
+        try {
+            FilePath parent = new FilePath(tmp);
```
Comment (MarkEWaite) [inline core/src/test/java/hudson/FilePathTest.java]: The coverage report says that the `return false` line is not covered.  It was also not reached when I ran the test in the debugger.  Please adjust the test to show the null pointer exception before your change.

Also, if you use the variable `temp` it is already available in the test class as a TempDir that will be automatically deleted at the end of the test.

**Review by MarkEWaite — CHANGES_REQUESTED** [review]

File: core/src/test/java/hudson/FilePathTest.java (hunk 5 above)
Comment (MarkEWaite) [reply core/src/test/java/hudson/FilePathTest.java]: I made several attempts to reach the added `return null` from a test.  I was unsuccessful.  I'm curious if you can find a way to reach it that I missed.

File: core/src/main/java/hudson/FilePath.java [hunk 6]
```diff
@@ -3781,6 +3787,11 @@ public Boolean invoke(@NonNull File parentFile, @NonNull VirtualChannel channel)
             Path currentFilePath = parentFile.toPath();
             while (!remainingPath.isEmpty()) {
                 Path directChild = this.getDirectChild(currentFilePath, remainingPath);
+                // --- FIX #2: Handle if directChild is null ---
+                if (directChild == null) {
+                    return false;
+                }
```
[the remark this answers is not in the document]
Comment (MarkEWaite) [reply core/src/main/java/hudson/FilePath.java]: > none appear to exercise the specific edge case where getDirectChild would return null (when the child path cannot be resolved to a direct child of the parent path).

I don't think that edge case can be exercised because there is specific code in the method that rejects absolute paths

File: core/src/main/java/hudson/FilePath.java [hunk 7]
```diff
@@ -3823,11 +3823,14 @@ public Boolean invoke(@NonNull File parentFile, @NonNull VirtualChannel channel)
...
         }
 
-        private @CheckForNull Path getDirectChild(Path parentPath, String childPath) {
+        private @NonNull Path getDirectChild(Path parentPath, String childPath) {
             Path current = parentPath.resolve(childPath);
             while (current != null && !parentPath.equals(current.getParent())) {
                 current = current.getParent();
             }
+            if (current == null) {
+                throw new IllegalStateException("Invalid path traversal: " + parentPath + " -> " + childPath);
```
Comment (daniel-beck) [inline core/src/main/java/hudson/FilePath.java:3885]: Please provide an example input where this could happen.

Comment (Vinamra-tech) [reply core/src/main/java/hudson/FilePath.java:3885]: @daniel-beck You are correct—there is no known input via the public API (isDescendant) that triggers this exception.

Context: I investigated this with @MarkEWaite, and we determined that the code path where current becomes null is effectively unreachable ("dead code") in the current implementation.

Why this change exists: We updated the annotation to @NonNull to fix a SpotBugs false positive. However, SpotBugs then flagged a new error because the method structure theoretically allows a null return (even if practically impossible).

This IllegalStateException was added to:

1. Satisfy the @NonNull contract for the static analyzer (proving it never returns null).

2. Provide a defensive guard rail so that if a future regression ever exposes this path, it fails with a clear error message instead of an NPE.

Comment (Vinamra-tech) [reply core/src/main/java/hudson/FilePath.java:3885]: @daniel-beck Also after forcing the state , through various means, I found out that I can trigger the exception if the method is called explicitly using Reflection to bypass the upstream logic and handling getDirectChild method directly. Since this is a possible way SpotBugs threw that false positive error and to satisfy it i did the mentioned changes.

Comment (Vinamra-tech) [reply core/src/main/java/hudson/FilePath.java:3885]: I apologise for dropping the comment and disturbing you through mail or other sort of notification, but it's been over a week since I dropped the comment for reviewing the pr could you please review it or tell me where I went wrong. Your feedback is much appreciated.

Comment (daniel-beck) [reply core/src/main/java/hudson/FilePath.java:3885]: Thanks for the ping. This should not have gone unanswered for a week.

> I can trigger the exception if the method is called explicitly using Reflection to bypass the upstream logic and handling getDirectChild method directly

What were the inputs for this?

Comment (Vinamra-tech) [reply core/src/main/java/hudson/FilePath.java:3885]: @daniel-beck No worries at all, thank you for getting back to me!

To trigger the exception via Reflection, I used a standard temporary directory as the parent and a child path containing excessive parent traversals (..) that go beyond the filesystem root.

The Inputs:

parent: A temporary directory path (e.g., /tmp/jenkins-test or similar valid Path).

child: A string with repeated .. segments (e.g., "../../../../../../../../../../").

The Execution Flow:

1. parentPath.resolve(childPath) creates a path that points "above" the root.

2. The while loop calls current.getParent() repeatedly.

3. Eventually, current reaches the filesystem root (e.g., /), and getParent() returns null.

4. The loop terminates because current is null, falling through to the if (current == null) check, which throws the IllegalStateException.

Comment (Vinamra-tech) [reply core/src/main/java/hudson/FilePath.java:3885]: @daniel-beck Sorry for dropping the comment on a weekend day , but I'm really cuious did you reproduce the test and the input we discussed previously or I'll have to come up with more solid test or reproduce the same test in my pr.

**Comment by Vinamra-tech** [comment]
> > Sorry for the ping but I am just curious if any maintainer tested my reflection test
> 
> Not yet. Many maintainers spent Friday at the Jenkins Contributor Summit in Belgium and are busy with other activities today and tomorrow. I don't expect any significant reviews until next week.

Could it now be reviewed, please?

**Comment by Vinamra-tech** [comment]
@daniel-beck Sorry for the ping but could the tests be verified?
Thank You.

**Comment by daniel-beck** [comment]
I plan to review this later this week.
